modules/3d/perf/perf_rendering.cpp

```diff
@@ -0,0 +1,227 @@
+    const std::array<std::string, 3> ShadingTypeEnum::svals{ std::string("White"),
+                                                             std::string("Flat"),
+                                                             std::string("Shaded")
+                                                           };
+
+    static inline void PrintTo(const ShadingTypeEnum &t, std::ostream *os) { t.PrintTo(os); }
+}
+
+enum class Outputs
+{
+    DepthOnly  = 0,
+    ColorOnly  = 1,
+    DepthColor = 2,
+};
+
+// that was easier than using CV_ENUM() macro
+namespace
+{
+    using namespace cv;
+    struct OutputsEnum
+    {
+        static const std::array<Outputs, 3> vals;
+        static const std::array<std::string, 3> svals;
+
+        OutputsEnum(Outputs v = Outputs::DepthColor) : val(v) {}
+        operator Outputs() const { return val; }
+        void PrintTo(std::ostream *os) const
+        {
+            int v = int(val);
+            if (v >= 0 && v < (int)vals.size())
+            {
+                *os << svals[v];
+            }
+            else
+            {
+                *os << "UNKNOWN";
+            }
+        }
+        static ::testing::internal::ParamGenerator<OutputsEnum> all()
+        {
+            return ::testing::Values(OutputsEnum(vals[0]),
+                                     OutputsEnum(vals[1]),
+                                     OutputsEnum(vals[2]));
+        }
+
+    private:
+        Outputs val;
+    };
+
+    const std::array<Outputs, 3> OutputsEnum::vals{ Outputs::DepthOnly,
+                                                    Outputs::ColorOnly,
+                                                    Outputs::DepthColor
+                                                  };
+    const std::array<std::string, 3> OutputsEnum::svals{ std::string("DepthOnly"),
+                                                         std::string("ColorOnly"),
+                                                         std::string("DepthColor")
+                                                       };
+
+    static inline void PrintTo(const OutputsEnum &t, std::ostream *os) { t.PrintTo(os); }
+}
+
+static Vec3f normalize_vector(Vec3f a)
+{
+    float length = std::sqrt(a[0] * a[0] + a[1] * a[1] + a[2] * a[2]);
+    return Vec3f(a[0] / length, a[1] / length, a[2] / length);
+}
+
+static Matx44f lookAtMatrixCal(const Vec3f& position, const Vec3f& lookat, const Vec3f& upVector)
+{
+    Vec3f w = normalize_vector(position - lookat);
+    Vec3f u = normalize_vector(upVector.cross(w));
+
+    Vec3f v = w.cross(u);
+
+    Matx44f res(u[0], u[1], u[2],   0,
+                v[0], v[1], v[2],   0,
+                w[0], w[1], w[2],   0,
+                   0,    0,    0,   1.f);
+
+    Matx44f translate(1.f,   0,   0, -position[0],
+                        0, 1.f,   0, -position[1],
+                        0,   0, 1.f, -position[2],
+                        0,   0,   0,          1.0f);
+    res = res * translate;
+
+    return res;
+}
+
```

Comment:
I propose to generalize the function and make it public API.

Reply:
Can be discussed; looks like a feature for a separate PR